File: src/zoho_books_cli/commands/recurring_invoices.py
"""`zb recurring-invoices ...` — full coverage of /recurringinvoices.

Thin wrappers: CRUD + update-by-custom-field + stop/resume + history (comments,
read-only) + per-recurring template apply. To list child invoices spawned
from a recurring invoice, filter the main invoices listing:
    zb invoices list --query recurring_invoice_id=<id>
Zoho does not expose a dedicated /recurringinvoices/{id}/childinvoices endpoint
(verified live: 404).
"""

from __future__ import annotations

import typer

from zoho_books_cli import config
from zoho_books_cli.client import ZohoBooksClient
from zoho_books_cli.commands import _shared

app = typer.Typer(
    help="Recurring invoice operations (CRUD + stop/resume + history + templates).",
    no_args_is_help=True,
)
templates_app = typer.Typer(help="Per-recurring-invoice template assignment.", no_args_is_help=True)
app.add_typer(templates_app, name="templates")

BASE = "/recurringinvoices"


@app.command("list")
def list_recurring_invoices(
    query: list[str] = typer.Option(
        None, "--query", "-q", help="Query params as key=value. May be repeated."
    ),
    params: str = typer.Option(
        None,
        "--params",
        help="Query params as a JSON object. Merged on top of --query.",
    ),
    page: int = typer.Option(None, "--page", help="Page number (1-indexed)."),
    per_page: int = typer.Option(None, "--per-page", help="Rows per page."),
    page_all: bool = typer.Option(
        False, "--page-all", help="Auto-paginate (NDJSON: one page per line)."
    ),
    page_limit: int = typer.Option(10, "--page-limit", help="Max pages with --page-all."),
    page_delay: int = typer.Option(
        100, "--page-delay", help="Delay between pages in ms with --page-all."
    ),
):
    """List recurring invoices. Returns one page plus page_context."""
    q = _shared.parse_query_pairs(query, params)
    if page is not None:
        q["page"] = str(page)
    if per_page is not None:
        q["per_page"] = str(per_page)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        _shared.emit_list_paginated(
            client,
            BASE,
            q,
            "recurring_invoices",
            page_all=page_all,
            page_limit=page_limit,
            page_delay_ms=page_delay,
        )


@app.command("create")
def create_recurring_invoice(
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Create a recurring invoice."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(BASE, json_body=json_body)
    _shared.emit_object(resp)


@app.command("get")
def get_recurring_invoice(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
):
    """Get a single recurring invoice by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{recurring_invoice_id}")
    _shared.emit_object(resp)


@app.command("update")
def update_recurring_invoice(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
    body: str = typer.Option(..., "--body", "-b", help="JSON body. IDs must be strings."),
):
    """Update a recurring invoice by ID."""
    json_body = _shared.parse_body(body)
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{recurring_invoice_id}", json_body=json_body)
    _shared.emit_object(resp)


@app.command("update-by-custom-field")
def update_by_custom_field(
    key: str = typer.Option(..., "--key", help="Custom-field API name."),
    value: str = typer.Option(..., "--value", help="Custom-field value to match on."),
    body: str = typer.Option(
        ..., "--body", "-b", help="JSON body with the update fields. IDs must be strings."
    ),
    upsert: bool = typer.Option(
        False, "--upsert", help="Create a new recurring invoice if no match is found."
    ),
):
    """Update a recurring invoice by a custom field's unique value.

    Identifier is sent via X-Unique-Identifier-Key / X-Unique-Identifier-Value headers
    per Zoho's spec; the request body carries only the fields to update.
    """
    json_body = _shared.parse_body(body)
    headers = {"X-Unique-Identifier-Key": key, "X-Unique-Identifier-Value": value}
    if upsert:
        headers["X-Upsert"] = "true"
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(BASE, json_body=json_body, headers=headers)
    _shared.emit_object(resp)


@app.command("delete")
def delete_recurring_invoice(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
):
    """Delete a recurring invoice by ID."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.delete(f"{BASE}/{recurring_invoice_id}")
    _shared.emit_action("recurring_invoice_id", recurring_invoice_id, resp)


@app.command("stop")
def stop(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
):
    """Stop a recurring invoice (POST /recurringinvoices/{id}/status/stop)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{recurring_invoice_id}/status/stop")
    _shared.emit_action("recurring_invoice_id", recurring_invoice_id, resp)


@app.command("resume")
def resume(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
):
    """Resume a stopped recurring invoice (POST /recurringinvoices/{id}/status/resume)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.post(f"{BASE}/{recurring_invoice_id}/status/resume")
    _shared.emit_action("recurring_invoice_id", recurring_invoice_id, resp)


@app.command("history")
def list_history(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
):
    """List history / comments for a recurring invoice (read-only)."""
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.get(f"{BASE}/{recurring_invoice_id}/comments")
    _shared.emit_list(resp, "comments")


@templates_app.command("apply")
def apply_template(
    recurring_invoice_id: str = typer.Argument(..., help="Zoho Books recurring_invoice_id."),
    template_id: str = typer.Argument(..., help="Zoho Books template_id."),
):
    """Apply a PDF template to a recurring invoice.

    PUT /recurringinvoices/{id}/templates/{template_id}.
    """
    cfg = config.load()
    with ZohoBooksClient(cfg) as client:
        resp = client.put(f"{BASE}/{recurring_invoice_id}/templates/{template_id}")
    _shared.emit_action("recurring_invoice_id", recurring_invoice_id, resp)

File: src/commands/recurring_invoices.rs
//! `zb recurring-invoices` — CRUD + stop/resume + history + templates over
//! `/recurringinvoices`.

use clap::{Args, Subcommand};

use crate::cli::Ctx;
use crate::client::RequestOptions;
use crate::commands::common::{self, BodyArgs, CustomFieldUpdateArgs, ListArgs};
use crate::errors::Result;
use crate::shared::Query;

const BASE: &str = "/recurringinvoices";
const ID_FIELD: &str = "recurring_invoice_id";

#[derive(Args, Debug)]
pub struct Cmd {
    #[command(subcommand)]
    pub sub: Sub,
}

#[derive(Subcommand, Debug)]
pub enum Sub {
    /// List recurring invoices.
    List(ListArgs),
    /// Create a recurring invoice.
    Create(BodyArgs),
    /// Get a single recurring invoice by ID.
    Get(IdArgs),
    /// Update a recurring invoice by ID.
    Update(UpdateArgs),
    /// Update a recurring invoice by a custom field's unique value.
    #[command(name = "update-by-custom-field")]
    UpdateByCustomField(CustomFieldUpdateArgs),
    /// Delete a recurring invoice by ID.
    Delete(IdArgs),
    /// Stop a recurring invoice.
    Stop(IdArgs),
    /// Resume a stopped recurring invoice.
    Resume(IdArgs),
    /// List history / comments for a recurring invoice (read-only).
    History(IdArgs),
    /// Per-recurring-invoice template assignment.
    Templates(TemplatesCmd),
}

#[derive(Args, Debug)]
pub struct IdArgs {
    pub recurring_invoice_id: String,
}

#[derive(Args, Debug)]
pub struct UpdateArgs {
    pub recurring_invoice_id: String,
    #[command(flatten)]
    pub body: BodyArgs,
}

#[derive(Args, Debug)]
pub struct TemplatesCmd {
    #[command(subcommand)]
    pub sub: TemplatesSub,
}

#[derive(Subcommand, Debug)]
pub enum TemplatesSub {
    /// Apply a PDF template to a recurring invoice.
    Apply(ApplyArgs),
}

#[derive(Args, Debug)]
pub struct ApplyArgs {
    pub recurring_invoice_id: String,
    pub template_id: String,
}

pub fn run(cmd: Cmd, ctx: &mut Ctx) -> Result<()> {
    match cmd.sub {
        Sub::List(args) => common::list(ctx, BASE, &args, "recurring_invoices"),
        Sub::Create(args) => common::create(ctx, BASE, &args),
        Sub::Get(args) => common::get(ctx, &format!("{BASE}/{}", args.recurring_invoice_id)),
        Sub::Update(args) => common::update(
            ctx,
            &format!("{BASE}/{}", args.recurring_invoice_id),
            &args.body,
        ),
        Sub::UpdateByCustomField(args) => common::update_custom(ctx, BASE, &args),
        Sub::Delete(args) => {
            let path = format!("{BASE}/{}", args.recurring_invoice_id);
            common::delete(ctx, &path, ID_FIELD, &args.recurring_invoice_id)
        }
        Sub::Stop(args) => {
            let path = format!("{BASE}/{}/status/stop", args.recurring_invoice_id);
            common::action(ctx, &path, ID_FIELD, &args.recurring_invoice_id)
        }
        Sub::Resume(args) => {
            let path = format!("{BASE}/{}/status/resume", args.recurring_invoice_id);
            common::action(ctx, &path, ID_FIELD, &args.recurring_invoice_id)
        }
        Sub::History(args) => {
            let path = format!("{BASE}/{}/comments", args.recurring_invoice_id);
            let resp = ctx.client.get(&path, &Query::new())?;
            common::emit_list_flat(&resp, "comments", ctx)
        }
        Sub::Templates(t) => match t.sub {
            TemplatesSub::Apply(args) => {
                let path = format!(
                    "{BASE}/{}/templates/{}",
                    args.recurring_invoice_id, args.template_id
                );
                let resp = ctx.client.put(&path, RequestOptions::default())?;
                common::emit_action(ID_FIELD, &args.recurring_invoice_id, &resp, ctx)
            }
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cli::Ctx;

    #[test]
    fn list_targets_recurringinvoices() {
        let mut server = mockito::Server::new();
        let m = server
            .mock("GET", "/books/v3/recurringinvoices")
            .match_query(mockito::Matcher::Any)
            .with_status(200)
            .with_body(r#"{"recurring_invoices":[]}"#)
            .create();
        let mut ctx = Ctx::new_for_test(&server.url());
        run(
            Cmd {
                sub: Sub::List(ListArgs::default()),
            },
            &mut ctx,
        )
        .unwrap();
        m.assert();
    }
}
